sklearn/tree/_criterion.pyx

```diff
@@ -599,6 +600,109 @@ cdef class Gini(ClassificationCriterion):
         impurity_right[0] = gini_right / self.n_outputs
 
 
+cdef class HellingerDistance(ClassificationCriterion):
+    r"""Hellinger distance criterion.
+
+    This handles cases where the target is a classification taking values 0 and 1.
+
+    Let:
+        - k represents a node before split
+        - k1 and k2 represent the children nodes after split
+        - left_0 represents observations of class 0 in node k1
+        - right_0 represents observations of class 0 in node k2
+        - left_1 represents observations of class 1 in node k1
+        - right_1 represents observations of class 1 in node k2
+        - parent_0 represents observations of class 0 in node k
+        - parent_1 represents observations of class 1 in node k
+
+    then let
+
+        weight_k1 = (\sqrt{\frac{N_{left_0}}{N_{parent_0}}}-\sqrt{\frac{N_{left_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k1 node.
+
+        weight_k2 = (\sqrt{\frac{N_{right_0}}{N_{parent_0}}}-\sqrt{\frac{N_{right_1}}{N_{parent_1}}})^2
+
+    be the partial hellinger distance score for the k2 node.
+
+    Hellinger distance score is then defined as:
+
+        hellinger_distance = \sqrt{weight_k1+weight_k2}
+
+    Note: hellinger distance range is [0, \sqrt{2}]
+
+    """
+    cdef double node_impurity(self) nogil:
+        """Evaluate the impurity of the current node.
+        
+        Evaluate the Hellinger distance criterion as impurity of the current node,
+        i.e. the impurity of samples[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef:
+            double impurity_left = 0.0
+            double impurity_right = 0.0
+            double impurity_sum = 0.0
+
+        # get the partial hellinger score, separately per child
+        self.children_impurity(&impurity_left, &impurity_right)
+
+        # hellinger score is sqrt of children impurity sum
+        impurity_sum = impurity_right + impurity_left
+        if impurity_sum >= 0:
+            impurity_sum = sqrt(impurity_sum)
```

Comment:
I think we should make sure that `impurity_sum` is by design positive.

Reply:
it is by design positive, can track it in the code
I added this check just out of habit, like checking for zero before division

I guess in case there will some impossible corner case, it will be better to crash than give bad impurity result

would you suggest to remove this check `if impurity_sum >= 0:?`